src/Zynga/Framework/StorableObject/Collections/Map/V2/Base.hh

```diff
@@ -0,0 +1,188 @@
+;
+use
+  Zynga\Framework\StorableObject\Collections\Map\V2\Importer\Storable as StorableImporter
+;
+use
+  Zynga\Framework\StorableObject\Collections\V2\Interfaces\StorableMapCollection
+;
+use
+  Zynga\Framework\StorableObject\V1\Exceptions\OperationNotSupportedException
+;
+use Zynga\Framework\StorableObject\V1\Fields;
+use Zynga\Framework\StorableObject\V1\Fields\Generic as FieldsGeneric;
+use Zynga\Framework\StorableObject\V1\Interfaces\ExportInterface;
+use Zynga\Framework\StorableObject\V1\Interfaces\FieldsInterface;
+use Zynga\Framework\StorableObject\V1\Interfaces\ImportInterface;
+use Zynga\Framework\StorableObject\V1\Interfaces\StorableObjectInterface;
+use Zynga\Framework\Type\V1\Interfaces\TypeInterface;
+
+class Base<Tv> implements StorableMapCollection<Tv> {
+  /**
+   * Native Map only supports int keys or string keys
+   * https://docs.hhvm.com/hack/reference/class/HH.Map/
+   * We are forcing it to support only string as for data with
+   * int keys we can make use of StorableVector
+   */
+  private Map<string, Tv> $map;
+  private bool $isRequired;
+  private bool $isDefaultValue;
+
+  public function __construct(private classname<Tv> $valueType) {
+    $interfaces = class_implements($valueType);
+    if (array_key_exists(StorableObjectInterface::class, $interfaces) === false &&
+        array_key_exists(TypeInterface::class, $interfaces) === false) {
+      throw new OperationNotSupportedException(
+        'Collection only support the following types for Map values:'.
+        StorableObjectInterface::class.
+        ', '.
+        TypeInterface::class,
+      );
+    }
+
+    $this->isRequired = false;
+    $this->isDefaultValue = true;
+    $this->map = Map {};
+  }
+
+  public function clear(): bool {
+    $this->map->clear();
+    $this->isDefaultValue = true;
+    return true;
+  }
+
+  public function isEmpty(): bool {
+    if ($this->map->count() == 0) {
+      return true;
+    }
+    return false;
+  }
+
+  public function count(): int {
+    return $this->map->count();
+  }
+
+  public function items(): Map<string, Tv> {
+    return $this->map;
+  }
+
+  public function add(Pair<string, Tv> $pair): this {
+    $this->map->add($pair);
+    return $this;
+  }
+
+  public function set(string $k, Tv $v): this {
+    $this->map->set($k, $v);
+    return $this;
+  }
+
+  public function addAll(?Traversable<Pair<string, Tv>> $iterable): this {
+    $this->map->addAll($iterable);
+    return $this;
+  }
+
+  public function setAll(?KeyedTraversable<string, Tv> $iterable): this {
+    $this->map->setAll($iterable);
+    return $this;
+  }
+
+  public function remove(string $key): this {
+    $this->map->remove($key);
+    return $this;
+  }
+
+  public function removeKey(string $key): this {
+    $this->map->removeKey($key);
+    return $this;
+  }
+
+  public function get(string $key): ?Tv {
+
+    if ($this->map->containsKey($key) === true) {
+      return $this->map[$key];
+    }
+
+    return null;
```

Comment:
Why return a nullable type at all? Null may be a valid case for a collection (if we support nullable types). Why not throw an out of bounds exception here?

Reply:
For that we already have function at(). Just trying to follow Map functionality as close as possible.

https://docs.hhvm.com/hack/reference/class/HH.Map/get/